tiledb/ml/readers/_tensor_schema.py

```diff
@@ -307,38 +345,66 @@ def iter_tensors(
                 yield tuple(SparseData(coords, d, shape) for d in data)
 
+
+RaggedArray = Sequence[np.ndarray]
+
+
+class RaggedTensorSchema(BaseSparseTensorSchema[RaggedArray]):
+    """TensorSchema for reading sparse TileDB arrays as ragged Numpy arrays.
+
+    Each ragged array is represented as a sequence of 1D Numpy arrays of the same dtype
+    but (in general) different size. Each item of the ragged array contains the values of
+    a given field for a given key (i.e. value along the key dimension). The items of the
+    ragged array are ordered by the key dimension but the values within each item are not
+    (meaningfully) ordered.
+
+    Example:
+    - (key, field1, field2) cells:
+        {('a', 3, 2.8), ('b', 1, 0.2), ('a', 7, 3.2), ('c', 5, 6.1), ('c', 2, 0.5), ('a', 2, 1.9)}
+
+    - Ragged arrays:
+        - key: [np.array(['a', 'a', 'a']), np.array(['b']), np.array(['c', 'c'])]
+        - field1: [np.array([3, 7, 2]), np.array([1]), np.array([5, 2])]
+        - field2: [np.array([2.8, 3.2, 1.9]), np.array([0.2]), np.array([6.1, 0.5])]
+    """
+
+    def __init__(self, **kwargs: Any):
+        super().__init__(**kwargs)
+        if self.key_dim not in self._query_kwargs["dims"]:
+            self._query_kwargs["dims"] += (self.key_dim,)
+
     @property
-    def max_partition_weight(self) -> int:
-        try:
-            memory_budget = int(self._array._ctx_().config()["py.init_buffer_bytes"])
-        except KeyError:
-            memory_budget = 10 * 1024**2
+    def shape(self) -> Sequence[Optional[int]]:
+        return len(self.key_range), None
 
-        # Determine the bytes per (non-empty) cell for each field.
-        # - For fixed size fields, this is just the `dtype.itemsize` of the field.
-        # - For variable size fields, the best we can do is to estimate the average bytes
-        #   size. We also need to take into account the (fixed) offset buffer size per
-        #   cell (=8 bytes).
-        offset_itemsize = np.dtype(np.uint64).itemsize
-        attr_or_dim_dtype = self._array.schema.attr_or_dim_dtype
-        query = self._array.query(return_incomplete=True, **self._query_kwargs)
-        est_sizes = query.multi_index[:].estimated_result_sizes()
-        bytes_per_cell = []
-        for field, est_result_size in est_sizes.items():
-            if est_result_size.offsets_bytes == 0:
-                bytes_per_cell.append(attr_or_dim_dtype(field).itemsize)
-            else:
-                num_cells = est_result_size.offsets_bytes / offset_itemsize
-                avg_itemsize = est_result_size.data_bytes / num_cells
-                bytes_per_cell.append(max(avg_itemsize, offset_itemsize))
+    def iter_tensors(
+        self, key_ranges: Iterable[InclusiveRange[int, int]]
+    ) -> Union[Iterable[RaggedArray], Iterable[Sequence[RaggedArray]]]:
+        query = self.query
+        get_data = itemgetter(*self._fields)
+        key_dim = self.key_dim
+        for key_range in key_ranges:
```

Comment:
@gsakkis what's the worst case scenario for this loop? Is there any meaning in parallelising it?

Reply:
For each `key_range` this performs a TileDB query. TileDB already parallelizes by default to all available cores so (at least in theory) there's no benefit parallelizing over multiple queries. 

Additionally, the partitioning into `key_ranges` tries to make each `key_range` as large as possible (within the bounds of `sm.memory_budget` and `py.init_buffer_bytes` config values for dense and sparse arrays, respectively). Running multiple queries will require more memory (or tweaking the `max_partition_weight` logic).

Finally this question is not specific to `RaggedTensorSchema`; if we investigate query parallelization, it makes sense to do it for all `TensorSchema`s.

Reply:
Then we should probably investigate parallelisation for all _TensorSchemas_ in a future ticket. Thanks!